dart/common/Addon.h

```diff
@@ -0,0 +1,305 @@
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class Properties : public Extensible<Properties> { };
+
+  /// Use the PropertiesMixer class to easily create a Properties extension
+  /// from an existing class or struct.
+  template <class Mixin>
+  using PropertiesMixer = ExtensibleMixer<Properties, Mixin>;
+
+  /// Virtual destructor
+  virtual ~Addon() = default;
+
+  /// Clone this Addon into a new manager
+  virtual std::unique_ptr<Addon> cloneAddon(AddonManager* newManager) const = 0;
+
+  /// Set the State of this Addon. By default, this does nothing.
+  virtual void setAddonState(const std::unique_ptr<State>& otherState);
+
+  /// Get the State of this Addon. By default, this returns a nullptr which
+  /// implies that the Addon is stateless.
+  virtual const State* getAddonState() const;
+
+  /// Set the Properties of this Addon. By default, this does nothing.
+  virtual void setAddonProperties(
+      const std::unique_ptr<Properties>& someProperties);
+
+  /// Get the Properties of this Addon. By default, this returns a nullptr
+  /// which implies that the Addon has no properties.
+  virtual const Properties* getAddonProperties() const;
+
+  /// This function will be called if the user is attempting to delete the Addon
+  /// but not immediately replacing it with another Addon of the same type. The
+  /// incoming argument will point to the AddonManager that had been holding
+  /// this Addon.
+  ///
+  /// If your Addon is mandatory for the AddonManager type that is passed in
+  /// here, then you should perform error handling in this function, and you
+  /// should return false to indicate that the operation is not permitted. If
+  /// you return false, then the Addon will NOT be removed from its Manager.
+  ///
+  /// By default, this simply returns true.
+  virtual bool isOptional(AddonManager* oldManager);
+
+protected:
+
+  /// Constructor
+  ///
+  // We require the AddonManager argument in this constructor to make it clear
+  // to extensions that they must have an AddonManager argument in their
+  // constructors.
+  Addon(AddonManager* manager);
+
+  /// This function should be overriden if your Addon needs to do any special
+  /// handling when its AddonManager gets changed.
+  virtual void changeManager(AddonManager* newManager);
+};
+
+//==============================================================================
+/// AddonWithProtectedState generates implementations of the State managing
+/// functions for an Addon class.
+template <class Base, typename StateData,
+          class ManagerType = AddonManager,
+          void (*updateState)(Base*) = &detail::NoOp<Base*> >
+class AddonWithProtectedState : public Addon
```

Comment:
Yep, this is a CRTP class that takes care of boilerplate code for managing the State and cloning. You have your actual Addon class inherit this, and it takes care of setting everything up. Inheriting this class also enables you to use macros for setting and getting each field in your State/Properties (although it just occurred to me that the macros have only been implemented for Properties fields so far; I still need to implement the State versions).